src/servers/tracer.cc

```diff
@@ -242,4 +240,203 @@ TraceManager::TraceActivity(
       << "\",\"ns\":" << timestamp_ns << "}]}";
 }
 
+void
+TraceManager::TraceTensorActivity(
+    TRITONSERVER_InferenceTrace* trace,
+    TRITONSERVER_InferenceTraceActivity activity, const char* name,
+    TRITONSERVER_DataType datatype, const void* base, size_t byte_size,
+    const int64_t* shape, uint64_t dim_count,
+    TRITONSERVER_MemoryType memory_type, int64_t memory_type_id, void* userp)
+{
+  if ((activity != TRITONSERVER_TRACE_TENSOR_INPUT) &&
+      (activity != TRITONSERVER_TRACE_TENSOR_OUTPUT)) {
+    LOG_ERROR << "Unsupported activity: "
+              << TRITONSERVER_InferenceTraceActivityString(activity);
+    return;
+  }
+
+  if ((memory_type != TRITONSERVER_MEMORY_CPU) &&
+      (memory_type != TRITONSERVER_MEMORY_CPU_PINNED)) {
+    LOG_ERROR << "Unsupported memory type: "
+              << TRITONSERVER_MemoryTypeString(memory_type);
+    return;
+  }
+
+  void* buffer_base = const_cast<void*>(base);
+
+  uint64_t id;
+  LOG_TRITONSERVER_ERROR(
+      TRITONSERVER_InferenceTraceId(trace, &id), "getting trace id");
+
+  // The function may be called with different traces but the same 'userp',
+  // group the activity of the same trace together for more readable output.
+  auto ts = reinterpret_cast<TraceStreams*>(userp);
+  std::stringstream* ss = nullptr;
+  {
+    if (activity == TRITONSERVER_TRACE_REQUEST_START) {
```

Comment:
You are never going to get this activity in this function. Instead you should just see if ts->streams_[id] exists and if it doesn't you need to create it. Also, because we can now get this function called by concurrently by request/response you need to have the ts->mtx_ lock stay active until update to 'ss' in complete.

You need to make similar change to TraceActivity function above.